Use more accurate ELF flags on MIPS

smrobtzz opened the pull request:
Changes the MIPS ELF flags used for metadata objects to be closer to what LLVM uses so the linker doesn't complain

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

smrobtzz committed:
Use more accurate ELF flags on MIPS

workingjubilee commented:
> Changes the MIPS ELF flags used for metadata objects to be closer to what LLVM uses so the linker doesn't complain


@smrobtzz what linker were you using from what distro to build what crate such that the linker complained?

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -270,45 +270,43 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
+                }
             };
 
+            if !sess.target.options.features.contains("noabicalls") {
+                e_flags |= elf::EF_MIPS_CPIC;
+            }
```
If this is defining whether it is position independent code for the entire emitted artifact, then no, I do not intend to have rustc respect another LLVM "feature" that defines a necessarily global quality of compilation, much less a negative feature. Rust has a model of what a target feature is that conflicts with that.

Certainly, respecting new CLI inputs like this is certainly something I would prefer we decide separately from what is a mere fix.

smrobtzz replied:
Makes sense, I'll remove that check.

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -270,45 +270,43 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
                 "n32" => e_flags |= elf::EF_MIPS_ABI2,
                 "o32" => e_flags |= elf::EF_MIPS_ABI_O32,
-                _ => e_flags |= elf::EF_MIPS_ABI_O32,
+                _ => {
+                    if architecture == Architecture::Mips {
+                        e_flags |= elf::EF_MIPS_ABI_O32
+                    }
+                }
```
I do not like this continued guessing. Either
- do nothing on this branch
- explicitly recognize `"n64"` as a possible variant, add it to all the 64-bit MIPS targets, and panic if `llvm_abiname` is unset

workingjubilee replied:
I am at about the point where I think we should require the field on all MIPSy targets, just like it is required on RISCV (last I checked) and a similar-but-different field is required on x86. Especially for ABI-affecting fields, we generally make fields required and then. It's basically the same thing as the lowercase issue: if there is a sensible default for or normalization of the data, then we should probably address that during deserialization and emit a warning if the field was not specified.

Or just panic on the field being unspecified, either one.

workingjubilee replied:
I am okay with the 32-bit case, I think, being inferred, due to LLVM not supporting other alternative MIPS ABIs for 32-bit MIPS (though supposedly they exist)

workingjubilee commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -3567,7 +3567,14 @@ impl Target {
             "s390x" => (Architecture::S390x, None),
             "mips" | "mips32r6" => (Architecture::Mips, None),
-            "mips64" | "mips64r6" => (Architecture::Mips64, None),
+            "mips64" | "mips64r6" => (
+                if self.options.llvm_abiname.to_lowercase().as_str() == "n32" {
+                    Architecture::Mips64_N32
+                } else {
+                    Architecture::Mips64
+                },
+                None,
+            ),
```
We don't support any N32 targets so this change is not required, really.

But more puzzlingly, I'm not sure why we are spending the effort to be case-indifferent when we're not even case-indifferent for "mips64" vs. "MIPS64"? If we want to normalize the case then we should have done it elsewhere instead of complicating every time we check.

smrobtzz replied:
I'm using a custom target spec with n32, so I need this change most. Will remove the to_lowercase though

workingjubilee replied:
I see. That makes sense.

It would be slightly more productive if we had an actual N32 target as a skeleton for understanding what we are supposed to be supporting, even if it was extremely rudimentary.

jieyouxu replied:
Can you please leave a comment to that effect? If we don't actually have such a Tier 3 N32 target that would hit this branch at all

smrobtzz replied:
Done

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -270,45 +270,43 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
-        Architecture::Mips => {
-            let arch = match sess.target.options.cpu.as_ref() {
+        Architecture::Mips | Architecture::Mips64 | Architecture::Mips64_N32 => {
+            let mut e_flags = match sess.target.options.cpu.as_ref() {
                 "mips1" => elf::EF_MIPS_ARCH_1,
                 "mips2" => elf::EF_MIPS_ARCH_2,
                 "mips3" => elf::EF_MIPS_ARCH_3,
                 "mips4" => elf::EF_MIPS_ARCH_4,
                 "mips5" => elf::EF_MIPS_ARCH_5,
-                s if s.contains("r6") => elf::EF_MIPS_ARCH_32R6,
+                "mips32r2" => elf::EF_MIPS_ARCH_32R2,
+                "mips32r6" => elf::EF_MIPS_ARCH_32R6,
+                "mips64r2" => elf::EF_MIPS_ARCH_64R2,
+                "mips64r6" => elf::EF_MIPS_ARCH_64R6,
                 _ => elf::EF_MIPS_ARCH_32R2,
             };
```
Mmh. If we set `EF_MIPS_ARCH_32R2` for `Architecture::Mips64` then I'm pretty sure that all further compilation logic is, uh, I believe the technical term is "totally fucked"? Please do not allow invalid combinations, so split things along the 32-bit and 64-bit lines and use `panic!()` or `bug!()` appropriately. If this means you need to extract code shared across the different MIPS bitnesses and ABIs into another function then that's fine.

workingjubilee replied:
If you can find a DiagCtxt a few steps up the callstack and want to thread it down (or pass more data up from here?) to make a nicer error, that would be fine too, just don't bother supporting obviously incorrect pairs.

smrobtzz replied:
OK, will fix

workingjubilee reviewed (changes requested):
I notice a number of changes that seem to be acting on a matter of "thoroughness" or otherwise changing things that don't seem to **need** to be changed in order for the compiler to be used to its current range of ability for the existing targets we support. With the somewhat terse PR description that did not describe the actual errors you encountered, I can't tell if all of these changes are actually required, and some of them seem to allow new forms of being incorrect.

While we *are* allowed to arbitrarily trash compilation if the `target-spec.json` (or our own target definitions!) are incorrect, so we can corrupt our artifacts (or preferably fail to compile) if the `llvm_abiname` is set in an incorrect or clashing way? I would prefer it if the change was slightly more guarded against other forms of incorrectness like the `target.options.cpu` or `target.options.features`, which can be set by ordinary compiler users giving it input, last I checked.

smrobtzz commented:
Sorry, should've explained more. I'm using a custom target specification with this combination of options:
```
"arch": "mips64",
"cpu": "mips3",
"llvm-abiname": "n32",
"features": "+noabicalls",
```

With these options, Rust's metadata objects (not sure what they're called) are always ELF64 mips64r2 objects, whereas the code objects emitted by LLVM are ELF32 mips3 objects. So, linking an executable (or, making an rlib and trying to link that with something else) fails because the objects are obviously incompatible. 

My changes are attempting to be "thorough" because the criteria for what the GNU and LLVM linkers consider compatible objects is pretty unclear. They can complain about nan2008 not matching, the ABI not matching, the CPU not matching, etc. So, I figured that it would be at better (resulting in the least amount of linker complaints for whatever custom or builtin targets people may be using) to at least put *some* effort into making sure the ELF files emitted are close to what LLVM outputs, rather than (for mips64 at least) defaulting to what `mips64el-linux-gnuabi64-gcc foo.c -c` did, and assuming that's fine.

smrobtzz commented:
I've added better validation that should catch the obviously wrong combinations of CPU, architecture, ABI, and removed the explicit check for `noabicalls`

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -270,45 +270,50 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
-                "mips1" => elf::EF_MIPS_ARCH_1,
-                "mips2" => elf::EF_MIPS_ARCH_2,
+        Architecture::Mips | Architecture::Mips64 | Architecture::Mips64_N32 => {
+            let is_32bit = architecture == Architecture::Mips;
```
```suggestion
            // "N32" indicates an "ILP32" data model on a 64-bit MIPS CPU
            // like SPARC's "v8+", x86_64's "x32", or the watchOS "arm64_32".
            let is_32bit = architecture == Architecture::Mips;
```

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -270,45 +270,50 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
-                "o32" => e_flags |= elf::EF_MIPS_ABI_O32,
-                _ => e_flags |= elf::EF_MIPS_ABI_O32,
+            // If the ABI is explicitly given, use it, or default to O32 on 32-bit MIPS,
+            // which is the only option.
```
```suggestion
            // which is the only "true" 32-bit option that LLVM supports.
```

workingjubilee commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -270,45 +270,50 @@ pub(super) fn elf_os_abi(sess: &Session) -> u8 {
 
             if sess.target.options.relocation_model != RelocModel::Static {
-                e_flags |= elf::EF_MIPS_PIC;
+                e_flags |= elf::EF_MIPS_PIC | elf::EF_MIPS_CPIC;
```
So I believe this matter is slightly confusing and we should comment on it:

```suggestion
                // PIC means position-independent code. CPIC means "calls PIC".
                // CPIC was mutually exclusive with PIC according to
                // the SVR4 MIPS ABI https://refspecs.linuxfoundation.org/elf/mipsabi.pdf
                // and should have only appeared on static objects with dynamically calls.
                // At some point someone (GCC?) decided to set CPIC even for PIC.
                // Nowadays various things expect both set on the same object file
                // and may even error if you mix CPIC and non-CPIC object files,
                // despite that being the entire point of the CPIC ABI extension!
                // As we are in Rome, we do as the Romans do.
                e_flags |= elf::EF_MIPS_PIC | elf::EF_MIPS_CPIC;
```

workingjubilee reviewed (approved):
Things look good overall now. I have some notes that might be useful to reference for later revisions of this code. Please feel free to reword them as you prefer, especially if they need corrections for accuracy.

smrobtzz committed:
Apply suggestions from code review

Co-authored-by: Jubilee <workingjubilee@gmail.com>

workingjubilee commented:
Thanks!

Hopefully this fixes the thing you're trying to wrangle.

@bors r+

bors commented:
:pushpin: Commit 57941afb23ef59e956d2a5ef6604369543edf419 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 57941afb23ef59e956d2a5ef6604369543edf419 -->
<!-- homu: {"type":"Approved","sha":"57941afb23ef59e956d2a5ef6604369543edf419","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
